src/platform/orchestrator/src/MpiApi.cpp

```diff
@@ -0,0 +1,489 @@
+            throw std::runtime_error("Route already exists for method and uri");
+        }
+    }
+    else
+    {
+        m_routes[uri] = std::map<http::Method, Handler>();
+    }
+
+    m_routes[uri][method] = handler;
+}
+
+http::Response Router::HandleRequest(const http::Request& request)
+{
+    http::Response response;
+
+    if (m_routes.find(request.uri) != m_routes.end())
+    {
+        if (m_routes[request.uri].find(request.method) != m_routes[request.uri].end())
+        {
+            OsConfigLogInfo(PlatformLog::Get(), "Received request for uri '%s' method '%d'", request.uri.c_str(), (int)request.method);
+            // TODO: validate request (Host/Accept/Content-Type)
+            m_routes[request.uri][request.method](request, response);
+        }
+        else
+        {
+            OsConfigLogError(PlatformLog::Get(), "Invalid request method '%d' for uri '%s'", (int)request.method, request.uri.c_str());
+            response.SetStatus(http::StatusCode::NOT_FOUND);
+        }
+    }
+    else
+    {
+        OsConfigLogError(PlatformLog::Get(), "Invalid request for uri '%s'", request.uri.c_str());
+        response.SetStatus(http::StatusCode::NOT_FOUND);
+    }
+
+    return response;
+}
+
+void Server::Listen(Router& router)
+{
+    struct stat st;
+    if (stat(g_socketPrefix, &st) == -1)
+    {
+        // TODO: use modes for root user
+        mkdir(g_socketPrefix, 0700);
+    }
+
+    if (0 <= (sockfd = socket(AF_UNIX, SOCK_STREAM | SOCK_NONBLOCK, 0)))
+    {
+        memset(&addr, 0, sizeof(addr));
+        addr.sun_family = AF_UNIX;
+        strncpy(addr.sun_path, g_mpiSocket, sizeof(addr.sun_path) - 1);
+        socklen = sizeof(addr);
+
+        // Unlink socket if already in-use
+        unlink(g_mpiSocket);
+        if (bind(sockfd, (struct sockaddr*)&addr, socklen) == 0)
+        {
+            if (listen(sockfd, 5) == 0)
+            {
+                OsConfigLogInfo(PlatformLog::Get(), "Listening on socket: '%s'", g_mpiSocket);
+
+                m_worker = std::thread(&Server::Worker, std::ref(*this), std::ref(router));
+            }
+            else
+            {
+                OsConfigLogError(PlatformLog::Get(), "Failed to listen on IPC socket: '%s'", g_mpiSocket);
+            }
+        }
+        else
+        {
+            OsConfigLogError(PlatformLog::Get(), "Failed to bind IPC socket: '%s'", g_mpiSocket);
+        }
+    }
+    else
+    {
+        OsConfigLogError(PlatformLog::Get(), "Failed to create IPC socket: '%s'", g_mpiSocket);
+    }
+}
+
+void Server::Stop()
+{
+    m_exitSignal.set_value();
+    m_worker.join();
+
+    OsConfigLogInfo(PlatformLog::Get(), "Server stopped");
+
+    close(sockfd);
+    unlink(g_mpiSocket);
+}
+
+void Server::Worker(Server& server, Router& router)
+{
+    int connfd = -1;
+    std::future<void> exitSignal = server.m_exitSignal.get_future();
+
+    while (exitSignal.wait_for(std::chrono::milliseconds(1)) == std::future_status::timeout)
```

Comment:
>while (exitSignal.wait_for(std::chrono::milliseconds(1)) == std::future_status::timeout) [](http://example.com/codeflow?start=4&length=88)

Don't do this. Instead rely on the DoWork callback that will be periodically called - and do not block during that